### ISSUE TITLE
No new Leader is assigned after Leader dies

### ISSUE THREAD

**Issue by Zelldon** [zeebe-io/zeebe#4810]
**Describe the bug**
On the long running cluster (stable nodes) we have a benchmark running with the following settings:

 * 3 partition
 * 3 nodes
 * 3 repl. factor
 * 1 starter - starts 100 workflow instances per seconds
 * 6 workers - activate 120 workers max

We can see in the metrics that the node 2 died on 2020-06-20 10:00 and came ready again at 2020-06-22 09:30.
We can also verify this in the [logs](https://console.cloud.google.com/logs/viewer?interval=JUMP_TO_TIME&authuser=1&project=zeebe-io&minLogLevel=0&expandAll=false&timestamp=2020-06-20T08:00:00.000000000Z&customFacets=&limitCustomFacetWidth=true&advancedFilter=resource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22zeebe-io%22%0Aresource.labels.location%3D%22europe-west1-b%22%0Aresource.labels.cluster_name%3D%22zeebe-long-running%22%0Aresource.labels.namespace_name%3D%22long-running%22%0Alabels.k8s-pod%2Fapp%3D%22long-running-zeebe%22%0Alabels.k8s-pod%2Fapp_kubernetes_io%2Finstance%3D%22long-running%22%0Alabels.k8s-pod%2Fapp_kubernetes_io%2Fname%3D%22zeebe-cluster%22%0AjsonPayload.logger!%3D%22io.zeebe.broker.exporter.elasticsearch%22%0Alabels.%22k8s-pod%2Fstatefulset_kubernetes_io%2Fpod-name%22%3D%22long-running-zeebe-2%22&scrollTimestamp=2020-06-22T08:21:19.824334000Z&dateRangeUnbound=forwardInTime)

```
I 2020-06-20T07:21:56.659539Z Bootstrap Broker-2 [6/10]: cluster services 
I 2020-06-20T07:21:57.082580Z Initializing Spring DispatcherServlet 'dispatcherServlet' 
I 2020-06-20T07:21:57.083878Z Initializing Servlet 'dispatcherServlet' 
I 2020-06-20T07:21:57.093290Z Completed initialization in 9 ms 
D 2020-06-22T08:21:19.804473Z Bootstrap Broker-2 [6/10]: cluster services started in 176363144 ms
```
It take for the node two, two complete days to come back! It might due to #4664 or something which is related to #4158 
During this time the processing stops completely! It sounds like a bug we have fixed before https://github.com/zeebe-io/zeebe/pull/4451

![noprogress](https://user-images.githubusercontent.com/2758593/85614672-68844700-b65b-11ea-8db3-0d291281dd7c.png)

We can see int the metrics that there is no leader choosen during this time!
![leaders](https://user-images.githubusercontent.com/2758593/85614954-cfa1fb80-b65b-11ea-9ef7-3280cff66a19.png)

Since we have so high atomix log level it seems there is [nothing happing in the nodes](https://console.cloud.google.com/logs/viewer?interval=JUMP_TO_TIME&authuser=1&project=zeebe-io&minLogLevel=0&expandAll=false&timestamp=2020-06-20T08:00:00.000000000Z&customFacets=&limitCustomFacetWidth=true&advancedFilter=resource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22zeebe-io%22%0Aresource.labels.location%3D%22europe-west1-b%22%0Aresource.labels.cluster_name%3D%22zeebe-long-running%22%0Aresource.labels.namespace_name%3D%22long-running%22%0Alabels.k8s-pod%2Fapp%3D%22long-running-zeebe%22%0Alabels.k8s-pod%2Fapp_kubernetes_io%2Finstance%3D%22long-running%22%0Alabels.k8s-pod%2Fapp_kubernetes_io%2Fname%3D%22zeebe-cluster%22%0AjsonPayload.logger!%3D%22io.zeebe.broker.exporter.elasticsearch%22%0A&scrollTimestamp=2020-06-22T08:20:01.464196000Z&pinnedLogId=suhmu7tz2v62uszs6&pinnedLogTimestamp=2020-06-20T07:51:15.181259Z)

```
D 2020-06-20T07:51:15.181259Z Partition role transitioning from FOLLOWER to CANDIDATE 
D 2020-06-20T07:51:25.215208Z Partition role transitioning from CANDIDATE to FOLLOWER 
D 2020-06-20T07:51:44.834318Z Detected unhealthy components. The current health status of components: {Raft-2=HEALTHY, logStream=UNHEALTHY} 
E 2020-06-20T07:51:44.834770Z Partition-2 failed, marking it as unhealthy 
D 2020-06-20T08:10:41.926965Z Partition role transitioning from FOLLOWER to CANDIDATE 
D 2020-06-20T08:10:46.831298Z Partition role transitioning from CANDIDATE to FOLLOWER 
D 2020-06-22T08:19:50.813829Z Partition role transitioning from FOLLOWER to CANDIDATE 
D 2020-06-22T08:19:51.849798Z Partition role transitioning from CANDIDATE to FOLLOWER 
D 2020-06-22T08:19:56.052915Z Partition role transitioning from FOLLOWER to CANDIDATE 

```

**Side note:**

It seems that over the days the memory is growing and it seems GC can't release everything or it least we have a lot of long living objects or a memory leak.

![highmem](https://user-images.githubusercontent.com/2758593/85612489-3b369980-b659-11ea-9b8b-e49383015fca.png)

With no load the memory looks like this:

![noload](https://user-images.githubusercontent.com/2758593/85612520-425da780-b659-11ea-844f-bb65d7611b38.png)

**To Reproduce**
Just run the cluster for long time and restart node.

**Expected behavior**
Leader is immediately chosen.

**Environment:**
 - OS: k8s
 - Zeebe Version: SNAPSHOT
 - Configuration: elastic

**Comment by Zelldon** [comment]
Might be related to #4723

**Comment by Zelldon** [comment]
@npepinpe I mentioned that we fixed with #4451 this bug and I'm wondering why this happens again. I think I understand know why. With refactoring and migrating the snapshotting #4424 there was a left over in the ZeebePartition which closes the SnapshotStore https://github.com/zeebe-io/zeebe/blob/develop/broker/src/main/java/io/zeebe/broker/system/partitions/ZeebePartition.java#L589-L602 which is wrong.

The fix for #4451 registers a snapshot listener in the PassiveRole, which should be called on new snapshots to delete logs with gaps. The problem is that closing the store, will remove all listeners. Closing the store will happen on tearDown the current partition.

We have the following order:

 * transition to follower in raft  (start role)
 * add snapshot listener to store
 * call role listeners
 * tearDown old partition
 * close SnapshotStore -> clears listeners
 * install follower zeebe partition 

The effect is that we haven't the fix anymore since the listener will not be called.
